crates/swc_css_modules/src/css_modules.rs

```diff
@@ -0,0 +1,266 @@
+};
+use swc_css_parser::{
+    lexer::Lexer,
+    parser::{Parser, ParserConfig},
+};
+use swc_css_visit::{Fold, FoldWith};
+
+use crate::hash::StyleHash;
+
+#[derive(Default)]
+pub struct CssModuleComponent {
+    pub hash_prefix: String,
+    pub class_collect: HashSet<String>,
+    pub group_pseudo_ident: Vec<Vec<String>>,
+}
+
+#[derive(Default)]
+pub struct ChildCssModuleComponent {
+    pub selector_list: Option<SelectorList>,
+}
+
+impl CssModuleComponent {
+    ///
+    /// init css-module context
+    pub fn new(filepath: &str, content: &str) -> Self {
+        CssModuleComponent {
+            hash_prefix: StyleHash::generate_css_module_hash(filepath, content),
+            class_collect: Default::default(),
+            group_pseudo_ident: vec![],
+        }
+    }
+
+    pub fn convert_token_and_span(t: TokenAndSpan) -> Result<String, String> {
+        let res: String;
+        if let Token::Ident { value, .. } = t.token {
+            res = value.to_string();
+        } else if let Token::Hash { value, .. } = t.token {
+            res = value.to_string();
+        } else if let Token::AtKeyword { value, .. } = t.token {
+            res = value.to_string();
+        } else if let Token::WhiteSpace { value } = t.token {
+            res = value.to_string();
+        } else if let Token::Number { value, .. } = t.token {
+            res = value.to_string();
+        } else if let Token::Percentage { value, .. } = t.token {
+            res = value.to_string();
+        } else if let Token::Url { value, .. } = t.token {
+            res = value.to_string();
+        } else if let Token::String { value, .. } = t.token {
+            res = value.to_string();
+        } else if let Token::BadString { value, .. } = t.token {
+            res = value.to_string();
+        } else if let Token::Delim { value, .. } = t.token {
+            res = value.to_string();
+        } else {
+            return Err("convert token has no matching correct type".to_string());
+        }
+        if !res.is_empty() {
+            Ok(res)
+        } else {
+            Err("convert token has error".to_string())
+        }
+    }
+
+    pub fn child_select_compile(select_txt: &str) -> ComplexSelector {
+        let content = select_txt.to_string() + "{" + "}";
+        let config = ParserConfig {
+            ..Default::default()
+        };
+        let cm: Lrc<SourceMap> = Default::default();
+        let fm = cm.new_source_file(FileName::Custom("/root/index.css".to_string()), content);
+        let lexer = Lexer::new(SourceFileInput::from(&*fm), config);
+        let mut parser = Parser::new(lexer, config);
+        let stylesheet = parser.parse_all().unwrap();
+        let mut visitor = ChildCssModuleComponent {
+            selector_list: None,
+        };
+        stylesheet.fold_with(&mut visitor);
+        let selector_list = visitor.selector_list.unwrap();
+        selector_list.children.get(0).unwrap().to_owned()
+    }
```

Comment:
It is very bad for perf, we should never do it

Reply:
sorry I don't know if swc can  provide me with a method to parse the string in global () directly into ast instead of str

Reply:
:global(.a)  -> complex_selector

Reply:
@zhusjfaker We just add the `css_modules` option, I can do it today